packages/material-ui-system/src/createStyled.js

```diff
@@ -52,8 +52,15 @@ const variantsResolver = (props, styles, theme, name) => {
   return variantsStyles;
 };
 
-export const shouldForwardProp = (prop) =>
-  prop !== 'styleProps' && prop !== 'theme' && prop !== 'sx' && prop !== 'as';
+export const shouldForwardProp = (prop) => {
+  return (
+    prop !== 'styleProps' &&
+    prop !== 'theme' &&
+    prop !== 'sx' &&
+    prop !== 'as' &&
+    prop !== 'componentState'
```

Comment:
Actually, can we follow emotion's approach and do not forward any custom props if the styled element is an HTML element? (cc @mnajdova)

Reply:
> Actually, can we follow emotion's approach and do not forward any custom props if the styled element is an HTML element? (cc @mnajdova)

Yeah, we could.

---

A bit background so that we can decide on `styleProps` vs `componentState`. When I initially introduced this prop, it was meant to only include props which are used in the styles. Then we saw the need for forwarding here more props (for example custom props for styling - maybe new color prop?). Currently the `styleProps` are implemented as all props + default values + state.

For the `styleProps`, the intention in the name looks better for me. If we name them `componentState`, then there will be question on, why some state is sent as a first class prop, and some is only in the `componentState`. Although the components are `unstyled` they should offer some injection point for applying styles, hence the name `styleProps`.

Anyway, the most important thing is that, we need a bag for all props so that we don't have to check every prop and decide whether it needs to be propagated to the DOM element or not.

Reply:
> Actually, can we follow emotion's approach and do not forward any custom props if the styled element is an HTML element?

@michaldudak Do you mean using `@emotion/is-prop-valid`? AFAIK, we are no using it. So far, we intercept the props we don't want to leak, one by one. It has a nice side effect to forces us to be more aware of how props flow. It might also have negative performance implications to use it. 

It seems that sc and em use it by default, while we don't. See the DOM output of: https://codesandbox.io/s/elegant-feather-5myf3?file=/src/App.js.

```jsx
import "./styles.css";
import styled1 from "styled-components";
import styled2 from "@emotion/styled";
import { styled as styled3 } from "@material-ui/system";

const Div1 = styled1("div")({});
const Div2 = styled2("div")({});
const Div3 = styled3("div")({});

export default function App() {
  return (
    <div className="App">
      <Div1 foo="foo">1</Div1>
      <Div2 foo="foo">2</Div2>
      <Div3 foo="foo">3</Div3>
    </div>
  );
}
```

<img width="278" alt="Capture d’écran 2021-06-29 à 12 57 45" src="https://user-images.githubusercontent.com/3165635/123786114-9ee14b80-d8d9-11eb-8c8e-ff616484bd34.png">

---

For the name `stateProps` could work too over `styleProps` and `componentState`.

Reply:
Yes, I meant `@emotion/is-prop-valid` (just for elements being HTML nodes). The performance hit should be neglible as they memoize the function call. But you're right that in most cases it shouldn't be necessary.

I think it's worth documenting that even though we are based on emotion, our behavior in this matter differs from theirs (BTW when I started working on this I used to spend some time and lost few hairs trying to figure out what's eating my `styleProps` prop... 😉).

As for the `styleProps`/`componentState` - after some thought I'm ok with passing all props and merging them with internal component state (although having event handlers there doesn't make much sense). 

Now, for the name - when I first saw it, I had hard time figuring out what `styleProps` are. Intuitively I thought it's a subset of props that's used for styling. First of all, they are technically not exactly props (they are host's props enriched with the computed state), so `stateProps` won't be my preference either. From the perspective of a receiving (slot) component, it's an object that describes the state of the host component. I suppose that in most cases it'll be indeed used to style the custom component, but it doesn't have to (nor should we care what it's used for). Perhaps `hostState` would suit you better, then?

Reply:
RFC created: https://github.com/mui-org/material-ui/issues/27127